src/mito2/src/memtable/merge_tree/data.rs

```diff
@@ -654,6 +659,73 @@ pub struct ParquetPart {
     data: Bytes,
 }
 
+/// Data parts under a shard.
+pub struct DataParts {
+    /// The active writing buffer.
+    pub(crate) active: DataBuffer,
+    /// immutable (encoded) parts.
+    pub(crate) frozen: Vec<DataPart>,
+}
+
+impl DataParts {
+    pub(crate) fn new(metadata: RegionMetadataRef, capacity: usize) -> Self {
+        Self {
+            active: DataBuffer::with_capacity(metadata, capacity),
+            frozen: Vec::new(),
+        }
+    }
+
+    /// Writes one row into active part.
+    pub fn write_row(&mut self, pk_id: PkId, kv: KeyValue) {
+        self.active.write_row(pk_id, kv)
+    }
+
+    /// Freezes the active data buffer into frozen data parts.
+    pub fn freeze(&mut self, pk_weights: &[u16]) -> Result<()> {
+        self.frozen.push(self.active.freeze(pk_weights)?);
+        Ok(())
+    }
+
+    /// Reads data from all parts including active and frozen parts.
+    /// The returned iterator yields a record batch of one primary key at a time.
+    /// The order of yielding primary keys is determined by provided weights.
+    pub fn read(&mut self, pk_weights: Vec<u16>) -> Result<DataPartsReader> {
+        let weights = Arc::new(pk_weights);
+        let mut nodes = Vec::with_capacity(self.frozen.len() + 1);
+        nodes.push(DataNode::new(DataSource::Buffer(
+            self.active.read(&weights)?,
```

Comment:
I realize that the weights to DataBuffer's read should be optional. If we are reading DataParts, the DataBuffer can also compare the pk index directly. But we can fix this later.

Reply:
> I realize that the weights to DataBuffer's read should be optional. If we are reading DataParts, the DataBuffer can also compare the pk index directly. But we can fix this later.

I added a todo [here](https://github.com/GreptimeTeam/greptimedb/pull/3361/commits/aba5c219492dc079ced984d7d0c87dc6647e2e6a#diff-5066fcdd3b887926278b1155cfbf005cbca27e9699e36292abcc55ecf80110e2R692) and I will fix this in next pr.